README.md
Raygun4PHP
==========

[Raygun.io](http://raygun.io) provider for PHP 5.3+

[![Build
Status](https://secure.travis-ci.org/MindscapeHQ/raygun4php.png?branch=master)](http://travis-ci.org/MindscapeHQ/raygun4php)

## Installation

Firstly, ensure that **curl** is installed and enabled in your server's php.ini file.

### With Composer

Composer is a package management tool for PHP which automatically fetches dependencies and also supports autoloading - this is a low-impact way to get Raygun4PHP into your site.

1. If you use a *nix environment, [follow the instructions](http://getcomposer.org/doc/01-basic-usage.md#installation) to install Composer. Windows users can run [this installer](https://github.com/johnstevenson/composer-setup) to automatically add it to the Path.

2. Inside your project's root directory create a composer.json file, containing:
```json
{
        "require": {
            "mindscape/raygun4php": "1.*"
        }
}
```
3. From your shell run `php composer.phar install` (*nix) or `composer install` (Windows). This will download Raygun4Php and create the appropriate autoload data.

4. Then in a PHP file just add:
```php
require_once 'vendor/autoload.php';
```
and the library will be imported ready for use.

### Manually with Git

Clone this repository and copy src/Raygun4php into an appropriate subdirectory in your project, such as /vendor/Raygun4php. Add `requires` definitions for RaygunClient.php where you want to make a call to Send().

```php
require (dirname(dirname(__FILE__)).'/vendor/Raygun4php/RaygunClient.php');
```
## Usage

You can send both PHP errors and object-oriented exceptions to Raygun. An easy way to accomplish this is to create a file containing exception and error handlers which make calls to the appropriate Raygun4PHP functions. As above, import Raygun4PHP - if you're using Composer, just add `require_once 'vendor/autoload.php'`, or if not manually import RaygunClient.php.

Then, create handlers that look something like this:

```php
namespace
{
	// paste your 'requires' statement

	$client = new \Raygun4php\RaygunClient("apikey for your application");

	function error_handler($errno, $errstr, $errfile, $errline ) {
		global $client;
  		$client->SendError($errno, $errstr, $errfile, $errline);
	}

	function exception_handler($exception)
	{
		global $client;
		$client->SendException($exception);
	}

	set_exception_handler('exception_handler');
	set_error_handler("error_handler");
}
```

Note that if you are placing in inside a file with a namespace of your choosing, the above code should be declared to be within the global namespace (thus the `namespace { }` is required). You will also need whichever `requires` statement as above (autoload or manual) before the `$client` instantiation.

Copy your application's API key from the Raygun.io dashboard, and place it in the constructor call as above (do not include the curly brackets).

If the handlers reside in their own file, just import it in every file where you'd like exceptions and errors to be sent, and they will be delivered to Raygun.io.

## Configuration

### Sending method - async/sync

Raygun4PHP has two algorithms which it can use to send your errors:

* **Asynchronous**: POSTs the message and returns to your script immediately without waiting for the response from the Raygun API.

* **Synchronous**: POSTs the message, blocks and receives the HTTP response from the Raygun API. This uses a socket connection which is still reasonably fast. This also allows the use of the debug mode to receive the HTTP response code; see below.


This can be set by passing in a boolean as the 2nd parameter to the constructor:

```php
$client = new \Raygun4php\RaygunClient("apiKey", $useAsyncSending);
```
#### $useAsyncSending options

Type: *boolean*

Linux/OS X default: *true*

Windows default: *false*

* If **$useAsyncSending** is *true*, and the script is running on a *nix platform, the message will be delivered asynchronously. SendError() and SendException() will return 0 if all went well.

* If **$useAsyncSending** is *false*, the script will block and receive the HTTP response.

*false* is the only effective option on Windows due to platform and library limitations within the supported versions.

### Proxies

A HTTP proxy can be set if your environment can't connect out through PHP or the `curl` binrary natively:

```php
$client = new \Raygun4php\RaygunClient("apiKey");
$client->setProxy('http://someproxy:8080');
```

### Debug mode

The client offers a debug mode in which the HTTP response code can be returned after a POST attempt. This can be useful when adding Raygun to your site. This is accessed by passing in *true* as the third parameter in the client constructor:

```php
$client = new \Raygun4php\RaygunClient("apiKey", $useAsyncSending, $debugMode);
```

#### $debugMode options

*Default: false*

If true is passed in, and **$useAsyncSending** is set to *false*, client->SendException() or SendError() will return the HTTP status code of the POST attempt.

**Note:** If $useAsyncSending is *true*, $debugMode is not available.

#### Response codes

* **202**: Message received by Raygun API correctly
* **403**: Invalid API key. Copy it from your Raygun Application Settings, it should be of the form `new RaygunClient("A+nUc2dLh27vbh8abls7==")`

### Version numbers

You can transmit the version number of your PHP project along with the message by calling `SetVersion()` on your RaygunClient after it is instantiated - this is optional but recommended as the version number is considered to be first-class data for a message.

### Affected user tracking

**New in 1.5: additional data support**

You can call $client->SetUser, passing in some or all of the following data, which will be used to provide an affected user count and reports:

```php
SetUser($user = null, $firstName = null, $fullName = null, $email = null, $isAnonymous = null, $uuid = null)
```

`$user` should be a unique identifier which is used to identify your users. If you set this to their email address, be sure to also set the $email parameter too.

This feature and values are optional if you wish to disable it for privacy concerns. To do so, pass `true` in as the third parameter to the RaygunClient constructor.

Note that this data is stored as cookies. If you do not call SetUser the default is to store a random UUID to represent the user.

This feature can be used in CLI mode by calling SetUser() at the start of your session.

### Filtering Sensitive Data

Some error data will be too sensitive to transmit to an external service, such as credit card details or passwords. Since this data is very application specific, Raygun doesn't filter out anything by default. You can configure to either replace or otherwise transform specific values based on their keys. These transformations apply to form data (`$_POST`), custom user data, HTTP headers, and environment data (`$_SERVER`). It does not filter the URL or its `$_GET` parameters, or custom message strings. Since Raygun doesn't log method arguments in stack traces, those don't need filtering. All key comparisons are case insensitive.

```php
$client = new \Raygun4php\RaygunClient("apiKey");
$client->setFilterParams(array(
	'password' => true,
	'creditcardnumber' => true,
	'ccv' => true,
	'php_auth_pw' => true, // filters basic auth from $_SERVER
));
// Example input: array('Username' => 'myuser','Password' => 'secret')
// Example output: array('Username' => 'myuser','Password' => '[filtered]')
```

You can also define keys as regular expressions:

```php
$client = new \Raygun4php\RaygunClient("apiKey");
$client->setFilterParams(array(
	'/^credit/i' => true,
));
// Example input: array('CreditCardNumber' => '4111111111111111','CreditCardCcv' => '123')
// Example output: array('CreditCardNumber' => '[filtered]','CreditCardCcv' => '[filtered]')
```

In case you want to retain some hints on the data rather than removing it completely, you can also apply custom transformations through PHP's anonymous functions. The following example truncates all keys starting with "address".

```php
$client = new \Raygun4php\RaygunClient("apiKey");
$client->setFilterParams(array(
	'Email' => function($key, $val) {return substr($val, 0, 5) . '...';}
));
// Example input: array('Email' => 'test@test.com')
// Example output: array('Email' => 'test@...')
```

Note that when any filters are defined, the Raygun error will no longer contain the raw HTTP data, since there's no effective way to filter it.

## Troubleshooting

As above, enable debug mode by instantiating the client with

```php
$client = new \Raygun4php\RaygunClient("apiKey", FALSE, TRUE);
```

This will echo the HTTP response code. Check the list above, and create an issue or contact us if you continue to have problems.

### 400 from command-line Posix environments

If, when running a PHP script from the command line on *nix operating systems, you receive a '400 Bad Request' error (when debug mode is enabled), check to see if you have any LESS_TERMCAP environment variables set. These are not compatible with the current version of Raygun4PHP. As a workaround, unset these variables before your script runs, then reset them afterwards.

## Changelog

- 1.6.0: Added HTTP proxy support, support X-Forwarded-For, null server var guards
- 1.5.3: Unify property casing (internal change)
- 1.5.2: Prevent error when query_string isn't present in $_SERVER
- 1.5.1: Guard against intermittent user id cookie being null; overload for disabling user tracking
- 1.5.0: Add enhanced user data support; fix null backtrace frames that could occur in 1.4
- 1.4.0: Added Sensitive Data Filtering; improved Error backtraces; Travis CI enabled
- 1.3.6: Move included Rhumsaa\Uuid lib into this namespace to prevent collisions if already included
- 1.3.5: Fixed possible bug in async curl logic
- 1.3.4: Bugfix in request message for testing
- 1.3.3: Hotfix for script error in v1.3.2
- 1.3.2: UTF-8 encoding routine from previous version updated to remove PHP 5.5 deprecated function
- 1.3.1: Request data, specifically $_SERVER variables, are now correctly encoded in UTF-8
- 1.3: Added debug mode to output HTTP response code when in socket mode
- 1.2.6: Fixed a bug in previous release rendering the UTC offset fix ineffective (thanks @mrardon for spotting this)
- 1.2.5: Request rawData (php://input) limited to 4096 bytes in line with other providers; clamp UTC offset to sane values as API was seeing some entries with max int offsets
- 1.2.4: Merged in unit tests
- 1.2.3: Fixed a bug where OccurredOn wasn't in correct ISO 8601 representation
- 1.2.2: Minor formatting refactor
- 1.2.1: Several bugfixes for user tracking and request processing
- 1.2: Added new async sending function; removed cURL dependency
- 1.1: Added user tracking support; improved experience in CLI mode; add user-specified timestamp support; fixed user data encoding error
- 1.0: Initial commit


CHANGELOG.md
- 1.7.0: Added SendRawException() for transmitting exception in logging scenario
- 1.6.0: Added HTTP proxy support, support X-Forwarded-For, null server var guards
- 1.5.3: Unify property casing (internal change)
- 1.5.2: Prevent error when query_string isn't present in $_SERVER
- 1.5.1: Guard against intermittent user id cookie being null; overload for disabling user tracking
- 1.5.0: Add enhanced user data support; fix null backtrace frames that could occur in 1.4
- 1.4.0: Added Sensitive Data Filtering; improved Error backtraces; Travis CI enabled
- 1.3.6: Move included Rhumsaa\Uuid lib into this namespace to prevent collisions if already included
- 1.3.5: Fixed possible bug in async curl logic
- 1.3.4: Bugfix in request message for testing
- 1.3.3: Hotfix for script error in v1.3.2
- 1.3.2: UTF-8 encoding routine from previous version updated to remove PHP 5.5 deprecated function
- 1.3.1: Request data, specifically $_SERVER variables, are now correctly encoded in UTF-8
- 1.3: Added debug mode to output HTTP response code when in socket mode
- 1.2.6: Fixed a bug in previous release rendering the UTC offset fix ineffective (thanks @mrardon for spotting this)
- 1.2.5: Request rawData (php://input) limited to 4096 bytes in line with other providers; clamp UTC offset to sane values as API was seeing some entries with max int offsets
- 1.2.4: Merged in unit tests
- 1.2.3: Fixed a bug where OccurredOn wasn't in correct ISO 8601 representation
- 1.2.2: Minor formatting refactor
- 1.2.1: Several bugfixes for user tracking and request processing
- 1.2: Added new async sending function; removed cURL dependency
- 1.1: Added user tracking support; improved experience in CLI mode; add user-specified timestamp support; fixed user data encoding error
- 1.0: Initial commit

src/Raygun4php/RaygunClient.php
<?php
namespace Raygun4php {
  require_once realpath(__DIR__ . '/RaygunMessage.php');
  require_once realpath(__DIR__ . '/RaygunIdentifier.php');
  require_once realpath(__DIR__ . '/Raygun4PhpException.php');
  require_once realpath(__DIR__ . '/Uuid.php');

  use Raygun4Php\Rhumsaa\Uuid\Uuid;

  class RaygunClient
  {
    protected $apiKey;
    protected $version;
    protected $tags;
    protected $user;
    protected $firstName;
    protected $fullName;
    protected $email;
    protected $isAnonymous;
    protected $uuid;
    protected $httpData;
    protected $useAsyncSending;
    protected $debugSending;
    protected $disableUserTracking;
    protected $disableDefaultFiltering;
    protected $proxy;

    /**
     * @var Array Parameter names to filter out of logged form data. Case insensitive.
     * Accepts regular expressions when the name starts with a forward slash.
     * Maps either to TRUE, or to a callable with $key and $value arguments.
     */
    protected $filterParams = array();

    private $host = 'api.raygun.io';
    private $path = '/entries';
    private $transport = 'ssl';
    private $port = 443;

    private $defaultFilterFields = array('PHP_AUTH_USER' => true, 'PHP_AUTH_PW' => true);

    /*
    * Creates a new RaygunClient instance.
    * @param bool $useAsyncSending If true, attempts to post rapidly and asynchronously the script by forking a cURL process.
    * RaygunClient cannot return the HTTP result when in async mode, however. If false, sends using a blocking socket connection.
    * This is the only method available on Windows.
    * @param bool $debugSending If true, and $useAsyncSending is true, this will output the HTTP response code from posting
    * error messages. See the GitHub documentation for code meaning. This param does nothing if useAsyncSending is set to true.
    */
    public function __construct(
      $key,
      $useAsyncSending = true,
      $debugSending = false,
      $disableUserTracking = false,
      $disableDefaultFiltering = false)
    {
      $this->apiKey = $key;
      $this->useAsyncSending = $useAsyncSending;
      $this->debugSending = $debugSending;
      $this->disableDefaultFiltering = $disableDefaultFiltering;

      if (!$disableUserTracking) {
        $this->SetUser();
      }

      $this->disableUserTracking = $disableUserTracking;
    }

    /*
    * Transmits an error to the Raygun.io API
    * @param int $errorno The error number
    * @param string $errstr The error string
    * @param string $errfile The file the error occurred in
    * @param int $errline The line the error occurred on
    * @param array $tags An optional array of string tags used to provide metadata for the message
    * @param array $userCustomData An optional associative array that can be used to place custom key-value
    * @param int $timestamp Current Unix timestamp in the local timezone, used to set when an error occurred.
    * data in the message payload
    * @return The HTTP status code of the result when transmitting the message to Raygun.io
    */
    public function SendError($errno, $errstr, $errfile, $errline, $tags = null, $userCustomData = null, $timestamp = null)
    {
      $message = $this->BuildMessage(new \ErrorException($errstr, $errno, 0, $errfile, $errline), $timestamp);

      if ($tags != null)
      {
        $this->AddTags($message, $tags);
      }

      if ($userCustomData != null)
      {
        $this->AddUserCustomData($message, $userCustomData);
      }

      return $this->Send($message);
    }

    /*
    * Transmits an exception to the Raygun.io API
    * @param \Exception $exception An exception object to transmit
    * @param array $tags An optional array of string tags used to provide metadata for the message
    * @param array $userCustomData An optional associative array that can be used to place custom key-value
    * data in the message payload
    * @param int $timestamp Current Unix timestamp in the local timezone, used to set when an exception occurred.
    * @return The HTTP status code of the result when transmitting the message to Raygun.io
    */
    public function SendException($exception, $tags = null, $userCustomData = null, $timestamp = null)
    {
      $message = $this->BuildMessage($exception, $timestamp);

      if ($tags != null)
      {
        $this->AddTags($message, $tags);
      }

      if ($userCustomData != null)
      {
        $this->AddUserCustomData($message, $userCustomData);
      }

      return $this->Send($message);
    }

    /*
     * Transmits a raw error as represented by an exception and originating location.
     * @param \Exception $exception The exception which will have its backtrace sent
     * @param string $file The originating file
     * @param int $line The originating line number
     * @param string $message The error message
     * @param string $className The class name of the error
     * @return The HTTP status code of the result when transmitting the message to Raygun.io
     */
    public function SendRawException($exception, $file, $line, $message, $className, $tags = null, $userCustomData = null, $timestamp = null)
    {
      $payload = $this->BuildMessageFromRaw($exception, $file, $line, $message, $className);

      if ($tags != null)
      {
        $this->AddTags($payload, $tags);
      }

      if ($userCustomData != null)
      {
        $this->AddUserCustomData($payload, $userCustomData);
      }

      return $this->Send($payload);
    }

    /*
     * Sets the version number of your project that will be transmitted
     * to Raygun.io.
     * @param string $version The version number in the form of x.x.x.x,
     * where x is a positive integer.
     *
     */
    public function SetVersion($version)
    {
      $this->version = $version;
    }

    /*
    *  Stores the current user of the calling application. This will be added to any messages sent
    *  by this provider. It is used in the dashboard to provide unique user tracking.
    *  If it is an email address, the user's Gravatar can be displayed. This method is optional,
    *  if it is not used a random identifier will be assigned to the current user.
    *  @param string $user A username, email address or other identifier for the current user
    *  of the calling application.
    *
    */
    public function SetUser($user = null, $firstName = null, $fullName = null, $email = null, $isAnonymous = null, $uuid = null)
    {
      $timestamp = time() + 60 * 60 * 24 * 30;

      $this->firstName = $this->StoreOrRetrieveUserCookie('rgfirstname', $firstName);
      $this->fullName = $this->StoreOrRetrieveUserCookie('rgfullname', $fullName);
      $this->email = $this->StoreOrRetrieveUserCookie('rgemail', $email);

      $this->uuid = $this->StoreOrRetrieveUserCookie('rguuidvalue', $uuid);
      $this->isAnonymous = $this->StoreOrRetrieveUserCookie('rgisanonymous', $isAnonymous ? 'true' : 'false') == 'true' ? true : false;

      if (is_string($user))
      {
        $this->user = $user;

        if (php_sapi_name() != 'cli' && !headers_sent())
        {
          setcookie('rguserid', $user, $timestamp);
          setcookie('rguuid', 'false', $timestamp);
        }
      }
      else
      {
        if (!array_key_exists('rguuid', $_COOKIE))
        {
          $this->user = (string)Uuid::uuid4();

          if (php_sapi_name() != 'cli' && !headers_sent())
          {
            setcookie('rguserid', $this->user, $timestamp);
            setcookie('rguuid', 'true', $timestamp);
          }
        }
        else if (array_key_exists('rguserid', $_COOKIE))
        {
          $this->user = $_COOKIE['rguserid'];
        }

        $this->isAnonymous = $this->StoreOrRetrieveUserCookie('rgisanonymous', 'true') == 'true';
      }
    }

    private function StoreOrRetrieveUserCookie($key, $value)
    {
      $timestamp = time() + 60 * 60 * 24 * 30;

      if (is_string($value))
      {
        if (php_sapi_name() != 'cli' && !headers_sent())
        {
          setcookie($key, $value, $timestamp);
        }

        return $value;
      }
      else
      {
        if (array_key_exists($key, $_COOKIE))
        {
          if ($_COOKIE[$key] != $value && php_sapi_name() != 'cli' && !headers_sent())
          {
            setcookie($key, $value, $timestamp);
          }
          return $_COOKIE[$key];
        }
      }

      return null;
    }

    /*
     * Sets a string array of tags relating to the message,
     * used for identification. These will be transmitted along with messages that
     * are sent.
     * @param array $tags The tags relating to your project's version
    */
    private function BuildMessage($errorException, $timestamp = null)
    {
      $message = new RaygunMessage($timestamp);
      $message->Build($errorException);

      $message = $this->BuildMessageInternal($message);

      return $message;
    }

    private function BuildMessageFromRaw($exception, $file, $line, $message, $className, $timestamp = null)
    {
      $raygunMessage = new RaygunMessage($timestamp);

      $raygunMessage->BuildFromRaw($exception, $file, $line, $message, $className);
      $raygunMessage = $this->BuildMessageInternal($raygunMessage);

      return $raygunMessage;
    }

    private function BuildMessageInternal($message)
    {
      $message->Details->Version = $this->version;
      $message->Details->Context = new RaygunIdentifier(session_id());

      if ($this->user != null)
      {
        $message->Details->User = new RaygunIdentifier($this->user, $this->firstName, $this->fullName, $this->email, $this->isAnonymous, $this->uuid);
      }
      else if (!$this->disableUserTracking && array_key_exists('rguserid', $_COOKIE))
      {
        $message->Details->User = new RaygunIdentifier($_COOKIE['rguserid']);
      }

      return $message;
    }

    private function AddTags(&$message, $tags)
    {
      if (is_array($tags))
      {
        $message->Details->Tags = $tags;
      }
      else
      {
        throw new \Raygun4php\Raygun4PhpException("Tags must be an array");
      }
    }

    private function AddUserCustomData(&$message, $userCustomData)
    {
      if ($this->is_assoc($userCustomData))
      {
        $message->Details->UserCustomData = $userCustomData;
      }
      else
      {
        throw new \Raygun4php\Raygun4PhpException("UserCustomData must be an associative array");
      }
    }

    private function is_assoc($array)
    {
      return (bool)count(array_filter(array_keys($array), 'is_string'));
    }

    /*
     * Transmits a RaygunMessage to the Raygun.io API. The default attempts to transmit asynchronously.
     * To disable this and transmit sync (blocking), pass false in as the 2nd parameter in RaygunClient's
     * constructor. This may be necessary on some Windows installations where the implementation is broken.
     * This is a lower level function used by SendException and SendError and one of those should be used preferrably.
     * @param Raygun4php\RaygunMessage $message A populated message to be posted to the Raygun API
     * @return The HTTP status code of the result after transmitting the message to Raygun.io
     * 202 if accepted, 403 if invalid JSON payload
     */
    public function Send($message)
    {
      if (empty($this->apiKey))
      {
        throw new \Raygun4php\Raygun4PhpException("API not valid, cannot send message to Raygun");
      }

      $message = $this->filterParamsFromMessage($message);
      $message = $this->toJsonRemoveUnicodeSequences($message);

      return $this->post($message, realpath(__DIR__ . '/cacert.crt'));
    }

    private function post($data_to_send, $cert_path)
    {
      $headers = 0;
      $remote = $this->transport . '://' . $this->host . ':' . $this->port;

      $context = stream_context_create();
      $result = stream_context_set_option($context, 'ssl', 'verify_host', true);

      if (!empty($cert_path))
      {
        $result = stream_context_set_option($context, 'ssl', 'cafile', $cert_path);
        $result = stream_context_set_option($context, 'ssl', 'verify_peer', true);
      }
      else
      {
        $result = stream_context_set_option($context, 'ssl', 'allow_self_signed', true);
      }

      if ($this->proxy) {
        $result = stream_context_set_option($context, 'http', 'proxy', $this->proxy);
      }

      if ($this->useAsyncSending && strtoupper(substr(PHP_OS, 0, 3)) !== 'WIN')
      {
        $curlOpts = array(
          "-X POST",
          "-H 'Content-Type: application/json'",
          "-H 'X-ApiKey: " . $this->apiKey . "'",
          "-d " . escapeshellarg($data_to_send),
          "--cacert '" . realpath(__DIR__ . '/cacert.crt') . "'"
        );
        if ($this->proxy) {
          $curlOpts[] = "--proxy '" . $this->proxy . "'";
        }
        $cmd = "curl " . implode(' ', $curlOpts) . " 'https://api.raygun.io:443/entries' > /dev/null 2>&1 &";
        exec($cmd, $output, $exit);
        return $exit;
      }
      else
      {
        $fp = stream_socket_client($remote, $err, $errstr, 10, STREAM_CLIENT_CONNECT, $context);

        if ($fp)
        {
          $req = '';
          $req .= "POST $this->path HTTP/1.1\r\n";
          $req .= "Host: $this->host\r\n";
          $req .= "X-ApiKey: " . $this->apiKey . "\r\n";
          $req .= 'Content-length: ' . strlen($data_to_send) . "\r\n";
          $req .= "Content-type: application/json\r\n";
          $req .= "Connection: close\r\n\r\n";
          fwrite($fp, $req);
          fwrite($fp, $data_to_send);

          $response = "";
          if ($this->debugSending)
          {
            while(!preg_match("/^HTTP\/[\d\.]* (\d{3})/", $response))
            {
              $response .= fgets($fp, 128);
            }

            fclose($fp);

            return $response;
          }
          else
          {
            fclose($fp);

            return 0;
          }
        }
        else
        {
          $errMsg = "<br/><br/>" . "<strong>Raygun Warning:</strong> Couldn't send asynchronously. ";
          $errMsg .= "Try calling new RaygunClient('apikey', FALSE); to use an alternate sending method, or RaygunClient('key', FALSE, TRUE) to echo the HTTP response" . "<br/><br/>";
          echo $errMsg;
          trigger_error('httpPost error: ' . $errstr);
          return null;
        }
      }
    }

    function toJsonRemoveUnicodeSequences($struct) {
      return preg_replace_callback("/\\\\u([a-f0-9]{4})/", function($matches){ return iconv('UCS-4LE','UTF-8',pack('V', hexdec("U$matches[1]"))); }, json_encode($struct));
    }

    /**
     * Optionally applies a value transformation to every matching key, as defined by {@link FilterParams}.
     * Replaces the value by default, but also supports custom transformations through
     * anonymous functions. Applies to form data, environment data, HTTP headers.
     * Does not apply to GET parameters in the request URI.
     * Filters out raw HTTP data in case any filters are defined, since we can't accurately filter it.
     *
     * @param RaygunMessage $message
     * @param  string $replace Value to be inserted by default (unless specified otherwise by custom transformations).
     * @return RaygunMessage
     */
    function filterParamsFromMessage($message, $replace = '[filtered]') {
      if (!$this->disableDefaultFiltering) {
        $filterParams = array_merge($this->getFilterParams(), $this->defaultFilterFields);
      } else {
        $filterParams = $this->getFilterParams();
      }

      // Skip checks if none are defined
      if(!$filterParams) {
        return $message;
      }

      // Ensure all filters are callable
      foreach($filterParams as $filterKey => $filterFn) {
        if(!is_callable($filterFn)) {
          $filterParams[$filterKey] = function($key, $val) use ($replace) {return $replace;};
        }
      }

      $walkFn = function(&$val, $key) use ($filterParams) {
        foreach($filterParams as $filterKey => $filterFn) {
          if(
            (strpos($filterKey, '/') === 0 && preg_match($filterKey, $key))
            || (strpos($filterKey, '/') === FALSE && strtolower($filterKey) == strtolower($key))
          ) {
            $val = $filterFn($key, $val);
          }
        }
      };

      if($message->Details->Request->Form) {
        array_walk_recursive($message->Details->Request->Form, $walkFn);
      }

      if($message->Details->Request->Headers) {
        array_walk_recursive($message->Details->Request->Headers, $walkFn);
      }

      if($message->Details->Request->Data) {
        array_walk_recursive($message->Details->Request->Data, $walkFn);
      }

      if($message->Details->UserCustomData) {
        array_walk_recursive($message->Details->UserCustomData, $walkFn);
      }

      // Unset raw HTTP data since we can't accurately filter it
      if($message->Details->Request->RawData && $this->getFilterParams()) {
        $message->Details->Request->RawData = null;
      }

      return $message;
    }

    /**
     * @param Array $params
     * @return \Raygun4php\RaygunClient
     */
    function setFilterParams($params) {
      $this->filterParams = $params;
      return $this;
    }

    /**
     * @return Array
     */
    function getFilterParams() {
      return $this->filterParams;
    }

    /**
     * Use a proxy for sending HTTP requests to Raygun.
     * @param String $proxy URL including protocol and an optional port, e.g. http://myproxy:8080
     * @return \Raygun4php\RaygunClient
     */
    function setProxy($proxy) {
      $this->proxy = $proxy;
      return $this;
    }

    /**
     * @return String
     */
    function getProxy() {
      return $this->proxy;
    }

    public function __destruct()
    {
      if ($this->httpData)
      {
        curl_close($this->httpData);
      }
    }

  }
}